### PR TITLE
Add dict(*iterable) constructor

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fix issue #8328

### CODE REVIEW

File: numba/typed/dictimpl.py [hunk 1]
```diff
@@ -1,9 +1,32 @@
...
 
 
+_message_dict_support = """
+Unsupported use of `dict()` with keyword argument(s). \
+The only supported uses are `dict()` or `dict(*iterable)`.
+""".strip()
+
+
+@lower_builtin(dict, types.IterableType)
+def list_constructor(context, builder, sig, args):
```
Comment (esc) [inline numba/typed/dictimpl.py:15]:
```suggestion
def dict_constructor(context, builder, sig, args):
```

File: numba/typed/dictimpl.py [hunk 2]
```diff
@@ -1,9 +1,32 @@
...
+    from numba.typed import Dict
+
+    dicttype = sig.return_type
+    kt, vt = dicttype.key_type, dicttype.value_type
+
+    def list_impl(iterable):
```
Comment (esc) [inline numba/typed/dictimpl.py:21]:
```suggestion
    def dict_impl(iterable):
```

File: numba/typed/dictimpl.py [hunk 3]
```diff
@@ -1,9 +1,32 @@
...
+        res = Dict.empty(kt, vt)
+        for k, v in iterable:
+            res[k] = v
+        return res
+
+    return context.compile_internal(builder, list_impl, sig, args)
```
Comment (esc) [inline numba/typed/dictimpl.py:27]:
```suggestion
    return context.compile_internal(builder, dict_impl, sig, args)
```

**Review by esc — CHANGES_REQUESTED** [review]
I gave this a first pass and found some potential copy and paste errors.

File: numba/typed/dictimpl.py (hunk 3 above)
Comment (guilhermeleobas) [reply numba/typed/dictimpl.py]: Fixed!

File: numba/core/typing/dictdecl.py [hunk 4]
```diff
@@ -17,13 +17,26 @@
 
 _message_dict_support = """
 Unsupported use of `dict()` with positional or keyword argument(s). \
-The only supported use is `dict()`.
+The only supported uses are `dict()` or `dict(*iterable)`.
```
Comment (stuartarchibald) [inline numba/core/typing/dictdecl.py:20]:
```suggestion
The only supported uses are `dict()` or `dict(iterable)`.
```

File: docs/source/reference/pysupported.rst [hunk 5]
```diff
@@ -724,10 +724,10 @@ Typed Dict
...
   ``dict()`` translates to calling ``numba.typed.Dict()``.
 
-Numba only supports the use of ``dict()`` without any arguments.  Such use is
-semantically equivalent to ``{}`` and ``numba.typed.Dict()``.  It will create
-an instance of ``numba.typed.Dict`` where the key-value types will be later
-inferred by usage.
+Numba supports the use of ``dict()``.  Such use is semantically equivalent to
+``{}`` and ``numba.typed.Dict()``. It will create an instance of
+``numba.typed.Dict`` where the key-value types will be later inferred by usage.
+Numba also supports ``dict(iterable)`` constructor as well.
```
Comment (stuartarchibald) [inline docs/source/reference/pysupported.rst:730]:
```suggestion
Numba also supports, explicitly, the ``dict(iterable)`` constructor.
```

File: numba/tests/test_dictobject.py [hunk 6]
```diff
@@ -1186,6 +1186,134 @@ def test_str(self):
         self.check_stringify(str)
 
 
+class DictIterableCtor(TestCase):
+
+    def setUp(self) -> None:
+        self.jit_enabled = None
+
+    def test_iterable_type_constructor(self):
```
Comment (stuartarchibald) [inline numba/tests/test_dictobject.py:1194]: This base class will run as a unit test, is that intended?

File: numba/tests/test_dicts.py [hunk 7]
```diff
@@ -37,6 +38,30 @@ def foo():
         d = foo()
         self.assertEqual(d, {1: 2})
 
+    def test_use_dict_iterable_args(self):
+        # Test dict(*iterable)
```
Comment (stuartarchibald) [inline numba/tests/test_dicts.py:42]:
```suggestion
        # Test dict(iterable)
```

File: numba/typed/typeddict.py [hunk 8]
```diff
@@ -117,6 +117,33 @@ def __init__(self, **kwargs):
...
+            # check if argument is iterable
+            arg = args[0]
+            if not isinstance(arg, Iterable):
+                msg = (f"'{type(arg)}' object is not iterable. Supported type "
+                       "constructor is Dict(iterable)")
+                raise TypeError(msg)
+
+            if isinstance(arg, Mapping):
+                for k, v in arg.items():
+                    self.__setitem__(k, v)
```
Comment (stuartarchibald) [inline numba/typed/typeddict.py:136]: This behaviour isn't working under JIT?

File: numba/tests/test_dictobject.py [hunk 9]
```diff
@@ -1186,6 +1186,134 @@ def test_str(self):
...
+        d = Dict({1: 2, 3: 4, 5: 6})
+        self.assertEqual(len(d), 3)
+        self.assertEqual(len(d.values()), 3)
+
+    def test_exception(self):
+        msg = '.*No implementation of function.*'
+        with self.assertRaisesRegex(TypingError, msg):
+            Dict({1: 2, 'a': 'b'})
+
+
```
Comment (stuartarchibald) [inline numba/tests/test_dictobject.py:1316]: Here's some things that either aren't working or could work better:
```python
from numba import njit
import numpy as np

# Valid, won't compile, fails at typing.
@njit
def foo():
    return dict(([1, 2], [1, 2]))

print(foo())


#Wrong error message, the issue is the types in the iterable aren't supported
#whereas the error message complains of: `Unsupported use of `dict()` with
#positional or keyword argument(s). The only supported uses are `dict()` or
#`dict(*iterable)`.`
@njit
def foo():
    return dict(((1, 2), (3, 'a')))

print(foo())


#Lowering error, should be caught earlier
@njit
def foo():
    return dict((('a', 'b', 'c'), ('d', 'e', 'f')))

print(foo())


#This should really complain about unhashable key type
@njit
def foo():
    return dict((('a', 1), ({}, 2)))

print(foo())


# Mapping support not working under JIT
@njit
def foo():
    return dict({x: x ** 2 for x in range(3)})

print(foo())
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @guilhermeleobas, I've not looked at the code in depth yet, but have provided some feedback about things that might need looking at in terms of usability and correctness. Thanks again!

File: numba/typed/typeddict.py (hunk 8 above)
Comment (guilhermeleobas) [reply numba/typed/typeddict.py]: I'll leave `dict(mapping)` ctor for a future PR and focus exclusively on `dict(iterable)` in this one.

File: numba/tests/test_dictobject.py (hunk 9 above)
Comment (guilhermeleobas) [reply numba/tests/test_dictobject.py]: I believe I have address all the issues pointed out in our comment, except for the last one, which involves `dict(mapping)`.

**Comment by guilhermeleobas** [comment]
Hi @stuartarchibald, I believe I have addressed all your comments. The current failure is not related to any changes introduced in my previous commit.

```
- Windows py310_np122

(testenv) D:\a\1\s>cmd /C conda install -q -y -c numba tbb=2021 tbb-devel 
Collecting package metadata (current_repodata.json): ...working... failed

CondaHTTPError: HTTP 504 GATEWAY TIMEOUT for url <[https://conda.anaconda.org/numba/win-64/repodata.json>](https://conda.anaconda.org/numba/win-64/repodata.json%3E)
Elapsed: 01:00.076791
CF-RAY: 73f5e5f559ed2ad6-ORD

A remote server error occurred when trying to retrieve this URL.
```

**Comment by esc** [comment]
@guilhermeleobas @kc611 -- may kindly request an update on the state of this PR?

**Comment by guilhermeleobas** [comment]
Hi @esc, PR is waiting for another review

File: numba/core/typing/dictdecl.py [hunk 10]
```diff
@@ -17,13 +17,39 @@
...
 class DictBuiltin(AbstractTemplate):
     def generic(self, args, kws):
-        if args or kws:
+        if kws:
             raise errors.TypingError(_message_dict_support)
+        if args:
+            iterable, = args
+            if isinstance(iterable, types.IterableType):
+                dtype = iterable.iterator_type.yield_type
+                if isinstance(dtype, types.UniTuple):
```
Comment (kc611) [inline numba/core/typing/dictdecl.py:33]: One question:

```python
@njit
def foo():
    return dict(([1, 2], [1, 2])) # Error

print(foo())
```

This is valid within CPython, But this PR only intends to add support when arguments are a list or a tuple of tuples (and not the other cases). Is that correct ?

**Review by kc611 — APPROVED** [review]
I gave it a review, its LGTM (so far). A question below.

File: numba/core/typing/dictdecl.py (hunk 10 above)
Comment (guilhermeleobas) [reply numba/core/typing/dictdecl.py:33]: If possible, I'd like to leave this use case for a future PR.

**Review by kc611 — APPROVED** [review]

**Review by sklam — APPROVED** [review]

**Pull request merged**